### DISCUSSION TITLE
User identity authentication in Azure Data Explorer and other Azure datasources

### DISCUSSION THREAD

**Discussion by kostrse** [grafana/grafana#62994] [Data Sources]
Authentication to Azure datasources (e.g. Azure Data Explorer) on behalf of the currently signed-in user is a popular customer request because of the way users used to work with these datasources for data exploration or due to security/audit policies of their organization.

To be able to authenticate to Azure services on behalf of a Grafana user, the user should be signed in to Grafana using Azure Active Directory (AAD) identity provider.

Then, the user's ID token. obtained during sign-in, could be combined with AAD client application credentials (client ID and secret) and be used to request the resource-specific access token.

Given that the ID token is already accessible to Grafana, there are two approaches how to provide configuration of the AAD application credentials.

## User identity with user-provided AAD configuration

This approach has been already implemented in Azure Data Explorer datasource (named as On-Behalf-Of) and assumes that Grafana users provide own AAD application credentials for token request via the datasource configuration UI:

<img width="659" alt="On-Behalf-Of authentication" src="https://user-images.githubusercontent.com/1131253/217156762-3b374605-51cc-4bff-835b-451448cc82f5.png">

## User identity with Grafana instance-level AAD configuration

Another proposed approach is to allow configuration of AAD application credentials by Grafana administrator or the cloud hosting environment, rather than end user.

In this case the `[azure]` section of Grafana config needs to be extended to allow configuration on on the level of Grafana instance:

```ini
[auth.azuread]
# Existing configuration section for sign-in via AAD
enabled = true

scopes = openid email profile

auth_url = https://login.microsoftonline.com/79de674a-fe58-4ab8-a714-545461651242/oauth2/v2.0/authorize
token_url = https://login.microsoftonline.com/79de674a-fe58-4ab8-a714-545461651242/oauth2/v2.0/token
client_id = c03d8fb1-987d-4c70-b3fb-31d3846ba9c4
client_secret = <client secret>

[azure]
managed_identity_enabled = false

# New configuration section for user identity token requests by datasources
user_identity_enabled = true

user_identity_token_url = https://login.microsoftonline.com/79de674a-fe58-4ab8-a714-545461651242/oauth2/v2.0/token
user_identity_client_id = c03d8fb1-987d-4c70-b3fb-31d3846ba9c4
user_identity_client_secret = <client secret>
```

The datasource configuration UI would not need any input form the user, similar to Managed Identity authentication:

<img width="659" alt="Current User authentication" src="https://user-images.githubusercontent.com/1131253/217156835-42b4db4b-4b7b-40c8-940a-914bc2ac536c.png">

The name "Current User" was chosen for simplicity, to communicate that this authentication works on behalf of the current user.

## Token provider implementation

Azure Data Explorer already has working implementation of on-behalf-of token provider ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/683000dd00e43e2fef3601b48d070b09d1340dfe/pkg/azuredx/adxauth/obo_token_provider.go)). The similar token provider needs to be implemented which would take AAD client credentials from `azsettings.AzureSettings` (from Grafana config), instead of `JsonData` ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/683000dd00e43e2fef3601b48d070b09d1340dfe/pkg/azuredx/adxauth/obo_token_provider.go#L32)).

### Alerts

User identity authentication works in context of the current user, therefore any background opeartions not associated with a specific user, like Alerts, are not supported. This is a trade off between convinience of configuration and limited functinality of user identity which can be communicated to users via intuitive UI and documentation.

Users should be adviced to use service identity when functionality like Alerts is necessary, and use user identity when convenience of data exploration is priority.

### Caching

If datasource implements data caching on backend, then the cache should be isolated by user since different users may have access to different parts of the data.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82694. Feel free to continue the discussion around this there. Thank you!
